### ISSUE TITLE
Advice for medaka_consensus speed increases for metagenomes

### ISSUE THREAD

**Issue by caity-s** [nanoporetech/medaka#49]
Hello,

Thank you for the recent update! The speed does seem to have improved, and I understand the program does take time, however I am wondering if there is anything I can do to process my metagenomes faster. There is a 24hr limit on our HPC cluster, and the processes aren't finishing in time before they are killed. It still seems to be the short region processing slowing things down, and only one thread is being used even though I specified to use all the available threads (24) on a ~700 GB RAM node. Is it possible for me to make use of more threads at this point?

The command I am currently using is:

`medaka_consensus -i $NPFASTQ -d draft_raconNP.fa -o medaka -t 24 -m r941_flip213`

And the timing:
```
[14:52:29 - Predict] Processing 28934 long region(s) with batching.
...
[19:37:00 - Predict] Processing 11274 short region(s).
...
[14:11:07 - Sampler] Pileup for tig00053666:0.0-7357.0 is of width 8279 ### short read processing still going the next day
```

I am testing it on a draft assembly (metagenome) that is 700 Mbp in size (all contigs >4 kbp, mean length 24kb), but this is our smallest assembly - the others are between 1-3 Gbp. Is there something I can do, such as in issue #39, like subsetting the assemblies, and recombining the split assembly at the end?

If the process is killed due to the 24 hr limit, can medaka pick up where it left off with the short reads if I restart the command?

Thank you for your help!

**Comment by cjw85** [comment]
Hi @caity-s,

The `medaka_consensus` program is good for simple datasets but perhaps not best for more complicated examples. If I explain what it does I suspect you will be able to find a better strategy.

`medaka_consensus` performs three tasks:

1. alignment or reads to input assembly (via `mini_align` which is a thin veil over `minimap2`)
2. running of consensus algorithm across assembly regions (via `medaka consensus`, note no underscore!)
3. aggregation of the results of 2. to create consensus sequences (`medaka stitch`)

The three steps are discrete, and you can run them yourself how you prefer. For splitting up your job it is Step 2. that you want to parallelize: the `medaka consensus` program can take a `--regions` argument which will restrict its action to particular assembly sequences from the `.bam` from Step 1. You can therefore submit individual jobs to cluster for each sequence (or small sets of sequences). In the final step, `medaka stitch` takes as input one or more of the `.hdf` files output by Step 2.

So in summary you can do something like:

```
# first align reads to assembly
mini_align -i basecalls.fasta -r assembly.fasta -P -m -p calls_to_draft.bam -t <threads>
# submit lots of jobs like this to your cluster, change model as appropriate:
mkdir results
medaka consensus calls_to_draft.bam results/contigs1-4.hdf \
  --model r941_flip235 --batch 200 --threads 8 \
  --region contig1 contig2 contig3 contig4
...
# wait for jobs, then collate results
medaka stitch results/*.hdf polished.assembly.fasta
```

I hope this helps.

**Comment by caity-s** [comment]
Hello @cjw85,

Thanks so much for that clear explanation - I will give your suggestions a go! Multiple batch jobs of regions will really help with the 24hr limit.

Cheers,
Caity

**Comment by caity-s** [comment]
Hello @cjw85,

Sorry, just a couple more questions:

1) If I set the threads of `medaka consensus` to -t 24 will it use more threads for the short regions than the one thread it uses when running through `medaka_consensus`? I notice that medaka processes the long regions "with batching" in the output, and is multithreaded, but this is not the case for the short regions. I am wondering why this is so?

2) I have 28000 contigs in the assembly, so that is a lot of jobs to submit, which our cluster might not be happy with (queuing systems and our node quotas etc) - when you say "small sets of sequences", can I do not-so-small batches of 3000 for example, submitting 10 jobs to the HPC (with a `--region <(cat contig_list_1_space_separated_on_one_line)`)? Or... am I limited only by the 2097152 character machine ARG_MAX....

Thank you so much,
Caity

**Comment by cjw85** [comment]
In answer to 1.: this is the expected behaviour, and somewhat of a technical annoyance. We can probably do better here, improving performance is something we might come back to after implementing some new features. I've made a internal issue to look at this.

For your second question, yes you will be limited by the maximum number of commandline args on your system; `medaka` will accept as many as you can give it. I only said "small" in the sense that the smaller the batches the more jobs you could submit to your cluster to spread the effort.

**Comment by caity-s** [comment]
Hello @cjw85,

Thank you for the clarification! And that's great to hear that multiprocessing of that small regions stage is on the cards for a future release - that will be really useful. For those interested I used `split -d -n l/20 racon_regions.txt split_regions- --additional-suffix .txt` on a regions file of the contig names (one per line). Then I replaced the new lines in the split files with spaces and did the following for all 20 of the `split_regions.txt` files:

```
medaka consensus $OUTDIR/temp/medaka1x/calls_to_draft.bam $OUTDIR/temp/medaka1x/split_regions-00.hdf --model $MODEL --batch 200 --threads 8 --region `cat $OUTDIR/temp/split_regions-00.txt`
```
I am currently running one as a test to see how long it takes on a slim node. Hopefully that will put me under the 24 hour limit. Just for interest, running `medaka_consensus` on the original file above finished yesterday after 4.5 days.

Thanks again,
Caity

**Comment by caity-s** [comment]
Hello @cjw85,

Split into 20 and using `medaka consensus` the samples took under an hour to complete (~30 mins per split file, which were all submitted at the same time). The sum of the 20 .hdf files is 40G. The .hdf file of the same sample processed using `medaka_consensus` is 80G. 

Should I be concerned that the .hdf files are so different? And that they are processed so fast (I was expecting 5ish hours per split file)? The log files all look fine.

Thanks,
Caity

**Comment by cjw85** [comment]
Hi @caity-s 

This does sound a little suspicious. If you run `medaka stitch` on the collection of 20 files it will index them and log all the contigs for which there is information in the files. If data is missing will also log that there are gaps. I would start by comparing the output of `medaka stitch` for your two runs.

**Comment by caity-s** [comment]
Hello @cjw85 ,

The logs seem to be similar - but not exactly the same. They were run on different servers (in-house server vs HPC). And, while medaka, samtools and minimap are the same versions, they did have slightly different racon versions so it is not a perfect match (v1.3.2 and 1.3.3):

medaka_consensus (racon v1.3.2, took 4.5 days on in-house server): 
The draft assembly (after racon, before medaka) = 700779337 bp
4 x `[22:45:28 - Stitch] tig00007173:71356.536-78044.0 ends before tig00007173:71356.1757-79230.0, skipping`
83 x `tig00524767:2.0-1421.0 and tig00524767:1603.0-9685.2 do not overlap`
83 x `[22:54:40 - Stitch] There is no overlap betwen tig00119401:34971.0-40561.0 and tig00119401:41151.0-48618.0`
28907 x `[22:40:44 - Stitch] Processing tig00000062`
Total basepairs of the consensus.fasta = 699802384

split medaka consensus (racon v1.3.3 - on HPC, and super fast):
The draft assembly (after racon, before medaka) = 700782455 bp
5 x `[09:16:17 - Stitch] tig00007173:71403.684-78301.0 ends before tig00007173:71403.1682-79286.0, skipping.`
73 x `tig00526448:5.0-6742.0 and tig00526448:7409.0-7475.0 do not overlap`
73 x  `[09:28:35 - Stitch] There is no overlap betwen tig00526799:3493.0-3645.0 and tig00526799:4774.0-9562.0`
28904 x `[09:11:19 - Stitch] Processing tig00000033.`
Total basepairs of the consensus.fasta = 699793208

All in all, it is tricky to say, I can't see any obvious errors. But it is suspiciously fast! I am currently trying to run an unsplit version of `medaka consensus` on the HPC (exact same racon input file), to see if that finishes faster too and creates the exact same output. If the split versions finished so fast, I expect it to finish in 10 hours unsplit using the same node structure...

Thanks,
Caity

**Comment by cjw85** [comment]
Given that your two `consensus.fasta` files are of equal length I would suggest everything has run OK. The consensus sequence is built up from scratch, not as edits to to the input sequence, such that if the `.hdf`s were missing data we'd expect the `consensus.fasta` sequence to be concomitantly shorter.

**Comment by caity-s** [comment]
That is really good to hear! Do you know why one hdf file is twice as big as the other (when the split hdf sizes are totalled)? That is weird considering there isn't twice as much sequence data stored within. 

Perhaps the speed is because of something as simple as node structure. Or maybe the splitting up of mini_align, consensus and stitch helped too. Would `medaka_consensus` call a different default of `medaka consensus` at all? 

Sorry for all the questions!

**Comment by caity-s** [comment]
Just a note: The exact same sample run on the same HPC node structure (same versions, node specs and racon assembly) without splitting did not complete within the 24 hour limit, and the hdf file was 47G and increasing. Splitting seems to make a big impact on time (and therefore node hour quotas!), and the .hdf size is not proportional to the splits for some reason.
